File: my_workshop/additions.py
"""Load external additions config for my-workshop.

The additions file lets users configure the SDK list, base image,
copy targets, and plug/slot connections.  Two locations are searched in order:

1. **Local** – ``workshop.my.yaml`` next to the resolved workshop YAML path.
2. **Global** – ``~/.config/my-workshop/my.yaml``.

When neither file exists, an empty dict is returned (noop).
See ``workshop.my.yaml.example`` for the full config format.
"""

import os

from .yaml_config import find_subblock, iter_sdk_blocks

# No built-in defaults — when no additions file is found the tool is a noop.
# Copy workshop.my.yaml.example and customise it for your project.
DEFAULT_ADDITIONS = {}


# ---------------------------------------------------------------------------
# File discovery
# ---------------------------------------------------------------------------

def find_additions(workshop_yaml_path):
    """Return the path to the additions file, or ``None``.

    *workshop_yaml_path* is the resolved path returned by ``find_yaml``.
    """
    # Local: same directory as the workshop YAML.
    local = os.path.join(os.path.dirname(workshop_yaml_path), "workshop.my.yaml")
    if os.path.exists(local):
        return local

    # Global: ~/.config/my-workshop/my.yaml
    home = os.environ.get("HOME", "")
    if home:
        global_path = os.path.join(home, ".config", "my-workshop", "my.yaml")
        if os.path.exists(global_path):
            return global_path

    return None


# ---------------------------------------------------------------------------
# Parsing
# ---------------------------------------------------------------------------

def _read_scalar(lines, key):
    """Return the value of a top-level ``key: value`` line, or ``None``."""
    prefix = key + ":"
    for line in lines:
        s = line.rstrip()
        if not s:
            continue
        if s.lstrip() == s and s.startswith(prefix):
            val = s[len(prefix):].strip()
            return val or None
    return None


def _subblock_range(lines, start, end, key):
    """Find ``(sub_start, sub_end)`` for ``key:`` between *start* and *end*.

    Returns ``None`` when *key* is not found.  ``sub_start`` is the line
    *after* the ``key:`` header; ``sub_end`` is one past the last line that
    belongs to the block (blank lines inside the block are included).
    """
    for i in range(start, end):
        s = lines[i].rstrip()
        if not s:
            continue
        indent = len(s) - len(s.lstrip())
        stripped = s.strip()
        if stripped == f"{key}:":
            sub_start = i + 1
            # Collect subsequent lines that are indented deeper than *key*.
            sub_end = sub_start
            for j in range(sub_start, end):
                sj = lines[j].rstrip()
                if not sj:
                    sub_end = j + 1
                    continue
                if len(sj) - len(sj.lstrip()) > indent:
                    sub_end = j + 1
                else:
                    break
            return sub_start, sub_end
    return None


def _parse_subblock_attrs(lines, header_idx):
    """Parse ``key: value`` pairs indented under the line at *header_idx*.

    Returns a dict of attributes.  The header line itself (e.g.
    ``my-plug:``) is at some indent; its children are one level deeper.
    """
    attrs = {}
    s = lines[header_idx].rstrip()
    base_indent = len(s) - len(s.lstrip())
    for j in range(header_idx + 1, len(lines)):
        sj = lines[j].rstrip()
        if not sj:
            continue
        indent = len(sj) - len(sj.lstrip())
        if indent <= base_indent:
            break
        # Only take direct children (one level deeper).
        if indent == base_indent + 2 and ":" in sj:
            k, _, v = sj.strip().partition(":")
            k = k.strip()
            v = v.strip()
            if k and v:
                attrs[k] = v
    return attrs


def _parse_list_items(lines, start, end, fields):
    """Yield dicts for YAML list items between *start* and *end*.

    *fields* is the set of scalar field names to extract (e.g.
    ``{"source", "target"}``).  Items are detected by ``- key: value`` lines
    at the first list-indent level, with subsequent ``key: value`` lines at
    the next level.
    """
    # Detect the base indent from the first non-empty line.
    base_indent = None
    for i in range(start, end):
        s = lines[i].rstrip()
        if s:
            base_indent = len(s) - len(s.lstrip())
            break
    if base_indent is None:
        return

    list_indent = base_indent          # indent of ``- ...``
    field_indent = base_indent + 2     # indent of subsequent ``key: val``
    item = None

    for i in range(start, end):
        s = lines[i].rstrip()
        if not s:
            continue
        indent = len(s) - len(s.lstrip())
        stripped = s.strip()

        if indent == list_indent and stripped.startswith("- "):
            # New list item.  Yield the previous one first.
            if item is not None:
                yield item
            item = {}
            # First field may sit on the same line as the dash.
            rest = stripped[2:].strip()
            if ":" in rest:
                k, _, v = rest.partition(":")
                k, v = k.strip(), v.strip()
                if k in fields and v:
                    item[k] = v
        elif indent == field_indent and item is not None and ":" in stripped:
            k, _, v = stripped.partition(":")
            k, v = k.strip(), v.strip()
            if k in fields and v:
                item[k] = v

    if item is not None:
        yield item


def parse_additions(text):
    """Parse additions config text into the standard dict shape.

    Returns ``{"base": str|None, "sdks": [...], "provision": {...}}``.
    Missing sections are filled with empty defaults.
    """
    lines = text.splitlines(keepends=True)
    result = {"base": None, "sdks": [], "provision": {"copy": [], "connect": []}}

    # --- base ---
    result["base"] = _read_scalar(lines, "base")

    # --- sdks ---  (reuse existing iter_sdk_blocks / find_subblock)
    for name, start, end in iter_sdk_blocks(lines):
        spec = {"name": name}
        for kind in ("plugs", "slots"):
            header, entries = find_subblock(lines, start, end, kind)
            if header is not None:
                block = {}
                # Find the range of the plugs/slots sub-block.
                bounds = _subblock_range(lines, start, end, kind)
                if bounds is not None:
                    sub_start, sub_end = bounds
                    for entry_name in entries:
                        # Find the entry's header line within the sub-block.
                        for i in range(sub_start, sub_end):
                            stripped = lines[i].rstrip().strip()
                            if stripped == f"{entry_name}:":
                                block[entry_name] = _parse_subblock_attrs(lines, i)
                                break
                spec[kind] = block
        result["sdks"].append(spec)

    # --- provision ---
    prov = _subblock_range(lines, 0, len(lines), "provision")
    if prov is not None:
        ps, pe = prov
        copy = _subblock_range(lines, ps, pe, "copy")
        if copy is not None:
            result["provision"]["copy"] = list(
                _parse_list_items(lines, copy[0], copy[1], {"source", "target"})
            )
        connect = _subblock_range(lines, ps, pe, "connect")
        if connect is not None:
            result["provision"]["connect"] = list(
                _parse_list_items(lines, connect[0], connect[1], {"plug", "slot"})
            )

    return result


# ---------------------------------------------------------------------------
# Loading
# ---------------------------------------------------------------------------

def load_additions(workshop_yaml_path):
    """Load the additions config, returning ``{}`` when no file is found."""
    path = find_additions(workshop_yaml_path)
    if path is None:
        return {}
    with open(path) as f:
        text = f.read()
    return parse_additions(text)

File: internal/additions/additions.go
// Package additions loads and parses optional external configuration for my-workshop.
package additions

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/bjornt/my-workshop/internal/workshop"
	"github.com/bjornt/my-workshop/internal/yamlconfig"
)

// Config holds optional additions to a workshop configuration.
// An empty/zero Config is a noop, matching the Python DEFAULT_ADDITIONS == {} semantics.
type Config struct {
	Base      string
	SDKs      []yamlconfig.SDKSpec
	Provision workshop.ProvisionSpec
}

// DefaultAdditions returns the empty/noop config.
func DefaultAdditions() Config {
	return Config{
		SDKs: []yamlconfig.SDKSpec{},
		Provision: workshop.ProvisionSpec{
			Copy:    []workshop.CopyEntry{},
			Connect: []workshop.ConnectEntry{},
		},
	}
}

// FindAdditions returns the path to the additions file, or "" if none exists.
// Preference: workshop.my.yaml next to workshopYAMLPath, then
// $HOME/.config/my-workshop/my.yaml when HOME is set.
func FindAdditions(workshopYAMLPath string) string {
	dir := filepath.Dir(workshopYAMLPath)
	local := filepath.Join(dir, "workshop.my.yaml")
	if _, err := os.Stat(local); err == nil {
		return local
	}

	home := os.Getenv("HOME")
	if home != "" {
		global := filepath.Join(home, ".config", "my-workshop", "my.yaml")
		if _, err := os.Stat(global); err == nil {
			return global
		}
	}

	return ""
}

// ParseAdditions parses additions config text into a Config.
// Missing sections are filled with empty defaults.
func ParseAdditions(text string) Config {
	lines := splitKeepEnds(text)
	result := DefaultAdditions()

	// --- base ---
	result.Base = readScalar(lines, "base")

	// --- sdks ---
	for _, block := range yamlconfig.IterSDKBlocks(lines) {
		spec := yamlconfig.SDKSpec{Name: block.Name}
		for _, kind := range []string{"plugs", "slots"} {
			header, entryNames := yamlconfig.FindSubblock(lines, block.Start, block.End, kind)
			if header == -1 {
				continue
			}
			entries := scanSDKEntries(lines, header+1, block.End, entryNames)
			if kind == "plugs" {
				spec.Plugs = entries
			} else {
				spec.Slots = entries
			}
		}
		result.SDKs = append(result.SDKs, spec)
	}

	// --- provision ---
	if provStart, provEnd, ok := subblockRange(lines, 0, len(lines), "provision"); ok {
		if copyStart, copyEnd, ok := subblockRange(lines, provStart, provEnd, "copy"); ok {
			for _, item := range parseListItems(lines, copyStart, copyEnd, map[string]bool{"source": true, "target": true}) {
				result.Provision.Copy = append(result.Provision.Copy, workshop.CopyEntry{
					Source: item["source"],
					Target: item["target"],
				})
			}
		}
		if connStart, connEnd, ok := subblockRange(lines, provStart, provEnd, "connect"); ok {
			for _, item := range parseListItems(lines, connStart, connEnd, map[string]bool{"plug": true, "slot": true}) {
				result.Provision.Connect = append(result.Provision.Connect, workshop.ConnectEntry{
					Plug: item["plug"],
					Slot: item["slot"],
				})
			}
		}
	}

	return result
}

// LoadAdditions loads the additions config for the given workshop YAML path.
// The returned bool is false when no additions file is found; the Config is
// then the empty/noop config.
func LoadAdditions(workshopYAMLPath string) (Config, bool) {
	path := FindAdditions(workshopYAMLPath)
	if path == "" {
		return DefaultAdditions(), false
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return DefaultAdditions(), false
	}
	return ParseAdditions(string(data)), true
}

// splitKeepEnds splits text into lines while keeping each trailing "\n".
// It mirrors Python's splitlines(keepends=True) for "\n"-delimited text.
func splitKeepEnds(text string) []string {
	var lines []string
	start := 0
	for i := range len(text) {
		if text[i] == '\n' {
			lines = append(lines, text[start:i+1])
			start = i + 1
		}
	}
	if start < len(text) {
		lines = append(lines, text[start:])
	}
	return lines
}

// readScalar returns the value of a top-level "key: value" line, or "" when absent.
func readScalar(lines []string, key string) string {
	prefix := key + ":"
	for _, line := range lines {
		s := strings.TrimRight(line, "\r\n")
		if s == "" {
			continue
		}
		if strings.TrimLeft(s, " ") == s && strings.HasPrefix(s, prefix) {
			val := strings.TrimSpace(s[len(prefix):])
			if val == "" {
				return ""
			}
			return val
		}
	}
	return ""
}

// subblockRange finds (start, end) for "key:" between start and end.
// start is the line after the key header; end is one past the last line that
// belongs to the block. It returns ok=false when key is not found.
func subblockRange(lines []string, start, end int, key string) (int, int, bool) {
	for i := start; i < end; i++ {
		s := strings.TrimRight(lines[i], "\r\n")
		if s == "" {
			continue
		}
		indent := len(s) - len(strings.TrimLeft(s, " "))
		stripped := strings.TrimSpace(s)
		if stripped == key+":" {
			subStart := i + 1
			subEnd := subStart
			for j := subStart; j < end; j++ {
				sj := strings.TrimRight(lines[j], "\r\n")
				if sj == "" {
					subEnd = j + 1
					continue
				}
				if len(sj)-len(strings.TrimLeft(sj, " ")) > indent {
					subEnd = j + 1
				} else {
					break
				}
			}
			return subStart, subEnd, true
		}
	}
	return 0, 0, false
}

// scanSDKEntries collects named plug/slot entries in file order between start
// and end, using the membership set returned by yamlconfig.FindSubblock.
func scanSDKEntries(lines []string, start, end int, names map[string]bool) yamlconfig.Entries {
	var entries yamlconfig.Entries
	for i := start; i < end; i++ {
		s := strings.TrimRight(lines[i], "\r\n")
		if s == "" {
			continue
		}
		indent := len(s) - len(strings.TrimLeft(s, " "))
		if indent <= 4 {
			break
		}
		stripped := strings.TrimSpace(s)
		if indent == 6 && strings.HasSuffix(stripped, ":") {
			name := strings.TrimSuffix(stripped, ":")
			if names[name] {
				entries = append(entries, yamlconfig.Entry{
					Name:  name,
					Attrs: parseSubblockAttrs(lines, i),
				})
			}
		}
	}
	return entries
}

// parseSubblockAttrs parses "key: value" pairs indented directly under the
// header line at headerIdx. Attribute order follows file order.
func parseSubblockAttrs(lines []string, headerIdx int) yamlconfig.Attrs {
	var attrs yamlconfig.Attrs
	s := strings.TrimRight(lines[headerIdx], "\r\n")
	baseIndent := len(s) - len(strings.TrimLeft(s, " "))
	for j := headerIdx + 1; j < len(lines); j++ {
		sj := strings.TrimRight(lines[j], "\r\n")
		if sj == "" {
			continue
		}
		indent := len(sj) - len(strings.TrimLeft(sj, " "))
		if indent <= baseIndent {
			break
		}
		if indent == baseIndent+2 && strings.Contains(sj, ":") {
			k, v, _ := strings.Cut(strings.TrimSpace(sj), ":")
			k = strings.TrimSpace(k)
			v = strings.TrimSpace(v)
			if k != "" && v != "" {
				attrs = append(attrs, yamlconfig.KV{Key: k, Value: v})
			}
		}
	}
	return attrs
}

// parseListItems yields dicts for YAML list items between start and end.
// Fields is the set of scalar field names to extract. Item order follows
// file order.
func parseListItems(lines []string, start, end int, fields map[string]bool) []map[string]string {
	var items []map[string]string
	baseIndent := -1
	for i := start; i < end; i++ {
		s := strings.TrimRight(lines[i], "\r\n")
		if s != "" {
			baseIndent = len(s) - len(strings.TrimLeft(s, " "))
			break
		}
	}
	if baseIndent == -1 {
		return items
	}

	listIndent := baseIndent
	fieldIndent := baseIndent + 2
	var item map[string]string

	for i := start; i < end; i++ {
		s := strings.TrimRight(lines[i], "\r\n")
		if s == "" {
			continue
		}
		indent := len(s) - len(strings.TrimLeft(s, " "))
		stripped := strings.TrimSpace(s)

		if indent == listIndent && strings.HasPrefix(stripped, "- ") {
			if item != nil {
				items = append(items, item)
			}
			item = make(map[string]string)
			rest := strings.TrimSpace(stripped[2:])
			if strings.Contains(rest, ":") {
				k, v, _ := strings.Cut(rest, ":")
				k = strings.TrimSpace(k)
				v = strings.TrimSpace(v)
				if fields[k] && v != "" {
					item[k] = v
				}
			}
		} else if indent == fieldIndent && item != nil && strings.Contains(stripped, ":") {
			k, v, _ := strings.Cut(stripped, ":")
			k = strings.TrimSpace(k)
			v = strings.TrimSpace(v)
			if fields[k] && v != "" {
				item[k] = v
			}
		}
	}
	if item != nil {
		items = append(items, item)
	}
	return items
}
